Add support for inherits and class references

bodgit opened the pull request:
Now detects violations such as:
```puppet
class foo inherits ::bar {}
```
and
```puppet
Class['::foo'] -> Class['::bar']
```
Fixes #20

bodgit committed:
Add support for inherits

Fixes #20

bodgit committed:
Add support for class references

alexjfisher commented:
Good stuff.  I'd actually written an implementation a few days back, but hadn't gotten around to committing it yet!  I prefer yours though. :)

alexjfisher commented on lib/puppet-lint/plugins/check_absolute_classname.rb:
```diff
@@ -40,6 +40,29 @@ def check
+        while s.type != :RBRACK
+          if s.type == :SSTRING && s.value.start_with?('::')
+            notify :warning, {
+              message: message,
```
We should probably use a different message here?
`class included by absolute name` isn't accurate.

alexjfisher reviewed (changes requested):
Looking good I think.  Could you update the README?

alexjfisher commented:
I can't decide whether the next release should be 3.0.0 or not.  I suppose this is 'breaking' in so much that lots of manifests that passed before, won't now.  'inherits' could be argued to be a way of including a class, but Class references aren't and checking them doesn't fall under the current gem's description (`A puppet-lint plugin to check that classes are included by their absolute name.`), which needs updating anyway.

@bodgit Could you update the gemspec with a new summary and description as well as updating the README?

ekohl commented:
I think this will work, but it's also valid to write `require => Class['::foo', '::bar']` and it should detect both. Another example would be `require => [Class['::foo'], Class['::bar']]`. Could you include tests for that as well?

bodgit committed:
Add some more tests

bodgit commented:
> I think this will work, but it's also valid to write `require => Class['::foo', '::bar']` and it should detect both. Another example would be `require => [Class['::foo'], Class['::bar']]`. Could you include tests for that as well?

I'd forgotten about the first syntax, thankfully the code still works with some extra tests added.

ekohl commented:
Thanks for adding those tests. If you address https://github.com/voxpupuli/puppet-lint-absolute_classname-check/pull/23#discussion_r506958924 I think this should be good to merge.

alexjfisher commented:
I guess bare words and interpolated strings are also valid puppet?  Ideally the plugin would also detect (and fix)

```puppet
Class[::foo] -> Class["::foo::install::${osfamily}"]
```

Not a blocker for me though.

bodgit commented:
Is it worth making the class references check separate within this gem?

alexjfisher commented:
> Is it worth making the class references check separate within this gem?

Possibly.  That would let users turn it off if they didn't like it.  Or we could default it to off for a release and then turn it on in the next major release.

alexjfisher commented:
But like I said in https://github.com/voxpupuli/puppet-lint-absolute_classname-check/pull/23#issuecomment-711166515, getting it perfect is not a blocker.  I think the new check is already going to be covering 90%+ of uses.

For the existing checking of how classes are included, I can keep coming up with obscure corner cases too, but life isn't long enough to try and catch them all.

eg (I've seen this style in the last week, complete with random trailing comma and all those single quotes)

```puppet
include '::foo::bar', '::foo:foobar', 'foo',
```

or as an actual array (instead of the comma separated string above)
```puppet
include [::foo::bar,'foo','::foo::foobar']
```

I think even the following is valid and would compile just fine...
```puppet
[
  '::foo::bar',
  'foo',
  ::foo::foobar,
].include
```

bodgit committed:
Split out class reference as separate check

bodgit committed:
Update README and gemspec

bodgit committed:
Support bare class names

bodgit commented:
I've split out the class reference check, updated the gemspec and README, and added support for bare class names as that was an easy addition. Anything else?

alexjfisher reviewed (approved):
@bodgit Superb contribution.  Thank you!
